### PR TITLE
Implement explicit tail calls

### PR DESCRIPTION
**Pull request by WaffleLapkin**
This is based on https://github.com/rust-lang/rfcs/pull/3407

r? @oli-obk 
cc @scottmcm @phi-go @DrMeepster

This is not complete, there are at least a couple of bugs (that I'm aware of, that is), however I'm starting to struggle because I've been looking at it too much, so a fresh look would be nice here.

I've tried to make commits as clean as I could, so hopefully rustfmt/clippy/libs/etc maintainers only need to look at a single commit that adds support there. (similarly all commits should pass tests at their point in time)

### CODE REVIEW

File: compiler/rustc_mir_build/src/check_tail_calls.rs [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+
+                    // Tail calling is likely to cause unrelated errors (ABI, argument mismatches)
+                    return;
+                }
+            };
+        }
+
+        // Erase regions since tail calls don't care about lifetimes
+        let callee_sig =
+            self.tcx.normalize_erasing_late_bound_regions(self.param_env, ty.fn_sig(self.tcx));
```
Comment (WaffleLapkin) [inline compiler/rustc_mir_build/src/check_tail_calls.rs:111]: `fn_sig` here panics on the following snippet:
```rust
fn g() {
    become (&g)();
}
```
(or in general any snippet that involves `&fn()` and such)

File: compiler/rustc_mir_build/src/check_tail_calls.rs [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        }
+
+        if caller_sig.inputs_and_output != callee_sig.inputs_and_output {
+            if caller_sig.inputs() != callee_sig.inputs() {
+                self.report_arguments_mismatch(expr.span, caller_sig, callee_sig);
+            }
+
+            if caller_sig.output() != callee_sig.output() {
+                span_bug!(expr.span, "hir typeck should have checked the return type already");
+            }
```
Comment (WaffleLapkin) [inline compiler/rustc_mir_build/src/check_tail_calls.rs:124]: This panics w/ opaque types, for example:
```rust
fn f() -> impl Sized {
    become g();
}

fn g() {}
```
or
```rust
fn f() -> impl std::future::Future<Output = ()> {
    become g();
}

async fn g() {}
```

both of those should probably work

**Review by JakobDegen — COMMENTED** [review]
I sort of started to review some of this but this is proving pretty difficult. I think the PR is significantly too big. There is no one person with the expertise to review all of this. The history is clean right now, but that will not be true after this inevitably goes through half a dozen rounds of review with a handful of people.

I think we should be looking to split this up. Each individual PR should be small enough that it can be reviewed in its entirety by one person, in such a way that after that review is done we're confident that either the code is correct, or there are actual issues tracking critical bugs that may not be forgotten about.

**Comment by RalfJung** [comment]
Agreed with @JakobDegen. Can this be split up into smaller PRs that only touch certain parts of the compiler -- the frontend, MIR building, codegen, interpreter, all separate? IIRC this is what we did with inline const.

There's no reason to have the Miri/interpreter or codegen implementation in the same PR that implements the checks enforcing the `become` restrictions, or the new dropck/borrowck interactions.

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 3]
```diff
@@ -646,6 +646,36 @@ pub enum TerminatorKind<'tcx> {
...
+    ///
+    /// [`Call`]: TerminatorKind::Call
+    /// [`Return`]: TerminatorKind::Return
+    TailCall {
+        /// The function that’s being called.
+        func: Operand<'tcx>,
+        /// Arguments the function is called with.
+        /// These are owned by the callee, which is free to modify them.
+        /// This allows the memory occupied by "by-value" arguments to be
+        /// reused across function calls without duplicating the contents.
```
[the remark this answers is not in the document]
Comment (WaffleLapkin) [reply compiler/rustc_middle/src/mir/syntax.rs:671]: This comment is copied from `TerminatorKind::Call::args`

**Pull request closed without merging**